### PR TITLE
INT B-19903 unlocking a move

### PR DESCRIPTION
**Pull request by danieljordan-caci**
## Summary

When setting up the functionality of an office user locking a move when they navigate into it, this sets the stage for when an office user needs to unlock that move. Currently there are two ways an office user can unlock a move:
1. Navigate to their respective queue (TOO -> `MovesQueue`, TIO -> `PaymentRequestQueue`, SC -> `ServicesCounselingQueue`)
2. Logout

This PR does the following:
- adds two service objects, `UnlockMove` and `CheckForLockedMovesAndUnlock` to be run for each scenario of an office user unlocking a move
- adds in logic checks to call `UnlockMove` when checks return `true` when loading up the queue list
- calls `checkForLockedMovesAndUnlock` when an office user logs out
- adds relevant tests

### How to test

This is a little tricky to test, so it is best to use two browsers that don't share cookies (I use one Chrome and one Safari - you could also do an incognito browser and one not)
1. Prior to firing up `server` & `client`, in your `envrc` change `FEATURE_FLAG_MOVE_LOCK=true`
2. Now run `make server_run` and `make client_run`
3. (OfficeUser1) Log in as an office user for each role and navigate into a move, but don't log out
4. (OfficeUser2) In your other browser log in as a different user and you should see these moves have a lock icon next to them in the queue
5. (OfficeUser1) Navigate back to your queue
6. (OfficeUser2) Refresh your screen - the lock icons should be gone
7. (OfficeUser1) Navigate back into a move
8. (OfficeUser2) Refresh your screen - lock icon should be shown next to the move
9. (OfficeUser1) Logout
10. (OfficeUser2) Refresh your screen - lock icon should be gone

### CODE REVIEW

File: pkg/handlers/authentication/auth.go [hunk 1]
```diff
@@ -475,6 +476,16 @@ func (h LogoutHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
 	if appCtx.Session() != nil {
+		// if the user is an office user, we need to unlock any moves that they have locked
+		if appCtx.Session().IsOfficeApp() && appCtx.Session().OfficeUserID != uuid.Nil {
+			moveUnlocker := movelocker.NewMoveUnlocker()
+			officeUserID := appCtx.Session().OfficeUserID
+			err := moveUnlocker.CheckForLockedMovesAndUnlock(appCtx, officeUserID)
+			if err != nil {
+				appCtx.Logger().Error("failed to unlock moves for office user")
+			}
+		}
```
Comment (cameroncaci) [inline pkg/handlers/authentication/auth.go:487]: Love this

File: pkg/handlers/authentication/auth.go [hunk 2]
```diff
@@ -475,6 +476,16 @@ func (h LogoutHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
 	}
 	if appCtx.Session() != nil {
+		// if the user is an office user, we need to unlock any moves that they have locked
+		if appCtx.Session().IsOfficeApp() && appCtx.Session().OfficeUserID != uuid.Nil {
+			moveUnlocker := movelocker.NewMoveUnlocker()
+			officeUserID := appCtx.Session().OfficeUserID
+			err := moveUnlocker.CheckForLockedMovesAndUnlock(appCtx, officeUserID)
+			if err != nil {
+				appCtx.Logger().Error("failed to unlock moves for office user")
+			}
```
Comment (cameroncaci) [inline pkg/handlers/authentication/auth.go:486]: Please add the `err` in the logged response as well as the officeUserID

File: pkg/handlers/ghcapi/queues.go [hunk 3]
```diff
@@ -76,6 +77,22 @@ func (h GetMovesQueueHandler) Handle(params queues.GetMovesQueueParams) middlewa
...
+					if lockedOfficeUserID != nil && *lockedOfficeUserID == officeUserID {
+						copyOfMove := move
+						unlockedMove, err := h.UnlockMove(appCtx, &copyOfMove, officeUserID)
+						if err != nil {
+							return queues.NewGetMovesQueueInternalServerError(), err
+						}
+						moves[i] = *unlockedMove
+					}
+				}
+			}
```
Comment (cameroncaci) [inline pkg/handlers/ghcapi/queues.go:94]: Clever way to catch it, but what if they're just utilizing multiple tabs? Think there's a frontend trigger we might be able to utilize?

File: pkg/handlers/authentication/auth.go (hunk 1 above)
Comment (danieljordan-caci) [reply pkg/handlers/authentication/auth.go:487]: LOL thanks

**Review by cameroncaci — CHANGES_REQUESTED** [review]
2 comments, 1 a request to add `err` and user to logger, and another to just ask a question

File: pkg/handlers/ghcapi/queues.go (hunk 3 above)
Comment (cameroncaci) [reply pkg/handlers/ghcapi/queues.go:94]: This one isn't a request for change, but a question

Comment (danieljordan-caci) [reply pkg/handlers/ghcapi/queues.go:94]: Not that I know of. Only UI trigger I know of is when state updates and that's handled by navigating/updating so I think throwing this in here is the right call.

File: pkg/handlers/authentication/auth.go (hunk 2 above)
Comment (danieljordan-caci) [reply pkg/handlers/authentication/auth.go]: Fixed

**Review by cameroncaci — APPROVED** [review]

**Pull request merged**